base/tps/src/main/java/org/dogtagpki/server/tps/rest/TPSCertService.java

```diff
@@ -162,6 +161,54 @@ public Response findCerts(String filter, String tokenID, Integer start, Integer
 
             return createOKResponse(response);
 
+        } catch (Exception e) {
+            throw new PKIException(e.getMessage());
+        }
+    }
+
+    private Response findAllCerts(String filter, Integer start, Integer size) {
+        String method = "TPSCertService:findAllCerts: ";
+        String msg = "";
+        TPSCertCollection response = new TPSCertCollection();
+
+        logger.debug("TPSCertService.findAllCerts({}, {}, {})", filter, start, size);
+
+        TPSEngine engine = TPSEngine.getInstance();
+        try {
+            TPSSubsystem subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+            TPSCertDatabase certDatabase = subsystem.getCertDatabase();
+            Iterator<TPSCertRecord> certRecs = certDatabase.findRecords(filter).iterator();
+            TokenDatabase tokenDatabase = subsystem.getTokenDatabase();
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+            if (authorizedProfiles == null) {
+                msg = "authorizedProfiles null";
+                logger.debug("{}{}", method, msg);
+                throw new PKIException(method + msg);
+            }
+            int total = 0;
+            while (certRecs.hasNext()) {
+                TPSCertRecord certRecord = certRecs.next();
+                String tokenID = certRecord.getTokenID();
+                TokenRecord tokenRecord = null;
+                try {
+                    tokenRecord = tokenDatabase.getRecord(tokenID);
+                } catch (Exception e) {
+                    // Proceed to next token if this one doesn't exist.
+                }
+                if (tokenRecord != null) {
+                    String type = tokenRecord.getType();
+                    if (type == null || type.isEmpty() || authorizedProfiles.contains(type) || authorizedProfiles.contains(UserResource.ALL_PROFILES)) {
```

Comment:
The `authorizedProfiles.contains(UserResource.ALL_PROFILES)` expression could be evaluated before the loop and stored into a boolean variable. If the value is `true` and assuming it's meant for the current page, the cert can be added into the results directly without retrieving the cert's token and checking the token type assuming, which could improve the performance.

Reply:
Sounds good, but I would prefer later, thanks!